File: Backend/inventory-service/src/main/java/com/sysco/inventory_service/controller/InventoryController.java
package com.sysco.inventory_service.controller;

import com.sysco.inventory_service.dto.CreateInventoryRequest;
import com.sysco.inventory_service.dto.InventoryDTO;
import com.sysco.inventory_service.dto.UpdateInventoryRequest;
import com.sysco.inventory_service.exception.InventoryAlreadyExistsException;
import com.sysco.inventory_service.exception.InventoryNotFoundException;
import com.sysco.inventory_service.service.InventoryService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import jakarta.validation.Valid;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/v1/inventory")
@RequiredArgsConstructor
@Slf4j
@CrossOrigin(origins = "*")
public class InventoryController extends InventoryAbstractController {

    private final InventoryService inventoryService;

    /**
     * Create inventory for a product - this is the main method to be called from product service
     */
    @PostMapping
    public ResponseEntity<Map<String, Object>> createInventoryForProduct(@Valid @RequestBody CreateInventoryRequest request) {
        try {
            InventoryDTO createdInventory = inventoryService.createInventoryForProduct(
                    request.getSku(), 
                    request.getQuantity()
            );
            return sendCreatedResponse(createdInventory, "Inventory created successfully for product SKU: " + request.getSku());
        } catch (InventoryAlreadyExistsException e) {
            log.error("Error creating inventory: {}", e.getMessage());
            return sendConflictResponse("Inventory already exists for product SKU: " + request.getSku());
        } catch (Exception e) {
            log.error("Unexpected error creating inventory: {}", e.getMessage());
            return sendInternalServerErrorResponse("Failed to create inventory due to an unexpected error");
        }
    }

    /**
     * Get inventory by SKU
     */
    @GetMapping("/{sku}")
    public ResponseEntity<Map<String, Object>> getInventoryBySku(@PathVariable String sku) {
        try {
            InventoryDTO inventory = inventoryService.getInventoryBySku(sku);
            return sendSuccessResponse(inventory, "Inventory retrieved successfully for SKU: " + sku);
        } catch (InventoryNotFoundException e) {
            log.error("Inventory not found: {}", e.getMessage());
            return sendNotFoundResponse("Inventory not found for SKU: " + sku);
        } catch (Exception e) {
            log.error("Unexpected error retrieving inventory: {}", e.getMessage());
            return sendInternalServerErrorResponse("Failed to retrieve inventory due to an unexpected error");
        }
    }

    /**
     * Update inventory quantity
     */
    @PutMapping("/{sku}")
    public ResponseEntity<Map<String, Object>> updateInventoryQuantity(
            @PathVariable String sku, 
            @Valid @RequestBody UpdateInventoryRequest request) {
        try {
            InventoryDTO updatedInventory = inventoryService.updateInventoryQuantity(sku, request.getQuantity());
            return sendSuccessResponse(updatedInventory, "Inventory quantity updated successfully for SKU: " + sku);
        } catch (InventoryNotFoundException e) {
            log.error("Inventory not found: {}", e.getMessage());
            return sendNotFoundResponse("Inventory not found for SKU: " + sku);
        } catch (Exception e) {
            log.error("Unexpected error updating inventory: {}", e.getMessage());
            return sendInternalServerErrorResponse("Failed to update inventory due to an unexpected error");
        }
    }

    /**
     * Check if inventory exists for a SKU
     */
    @GetMapping("/{sku}/exists")
    public ResponseEntity<Map<String, Object>> checkInventoryExists(@PathVariable String sku) {
        try {
            boolean exists = inventoryService.inventoryExists(sku);
            return sendSuccessResponse(Map.of("exists", exists), "Inventory existence check completed for SKU: " + sku);
        } catch (Exception e) {
            log.error("Unexpected error checking inventory existence: {}", e.getMessage());
            return sendInternalServerErrorResponse("Failed to check inventory existence due to an unexpected error");
        }
    }

    /**
     * Get all inventories
     */
    @GetMapping
    public ResponseEntity<Map<String, Object>> getAllInventories() {
        try {
            List<InventoryDTO> inventories = inventoryService.getAllInventories();
            return sendSuccessResponse(inventories, "All inventories retrieved successfully");
        } catch (Exception e) {
            log.error("Unexpected error retrieving all inventories: {}", e.getMessage());
            return sendInternalServerErrorResponse("Failed to retrieve inventories due to an unexpected error");
        }
    }
}

File: BFF/src/controllers/inventory.controller.js
const AbstractController = require('./AbstractController');
const inventoryService = require('../services/inventory.service');

class InventoryController extends AbstractController {

    /**
     * Create inventory for a product
     * POST /api/inventory
     */
    async createInventoryForProduct(req, res) {
        try {
            const createInventoryRequest = req.body;
            this.logRequest(req, 'Creating inventory for product');
            
            // Basic validation
            if (!createInventoryRequest.sku) {
                return this.sendBadRequestResponse(res, 'SKU is required');
            }

            if (!this.isValidSkuCode(createInventoryRequest.sku)) {
                return this.createValidationErrorResponse(res, 'SKU', 'must be between 2 and 50 characters');
            }

            if (createInventoryRequest.quantity === undefined || createInventoryRequest.quantity < 0) {
                return this.sendBadRequestResponse(res, 'Quantity must be a non-negative number');
            }

            const result = await inventoryService.createInventoryForProduct(createInventoryRequest);
            
            // If the service returns the standard format, pass it through
            if (result && result.status !== undefined && result.data !== undefined) {
                return res.status(result.status).json(result);
            }
            
            return this.sendCreatedResponse(res, result, `Inventory created successfully for product SKU: ${createInventoryRequest.sku}`);
        } catch (error) {
            console.error('[InventoryController] Error in createInventoryForProduct:', error.message);
            return this.handleAxiosError(res, error, 'Error creating inventory');
        }
    }

    /**
     * Get inventory by SKU
     * GET /api/inventory/:sku
     */
    async getInventoryBySku(req, res) {
        try {
            const { sku } = req.params;
            this.logRequest(req, `Getting inventory for SKU: ${sku}`);
            
            if (!this.isValidSkuCode(sku)) {
                return this.createValidationErrorResponse(res, 'SKU', 'must be between 2 and 50 characters');
            }

            const result = await inventoryService.getInventoryBySku(sku);
            
            // If the service returns the standard format, pass it through
            if (result && result.status !== undefined && result.data !== undefined) {
                return res.status(result.status).json(result);
            }
            
            return this.sendSuccessResponse(res, result, `Inventory retrieved successfully for SKU: ${sku}`);
        } catch (error) {
            console.error('[InventoryController] Error in getInventoryBySku:', error.message);
            return this.handleAxiosError(res, error, `Error retrieving inventory for SKU: ${req.params.sku}`);
        }
    }

    /**
     * Update inventory quantity
     * PUT /api/inventory/:sku
     */
    async updateInventoryQuantity(req, res) {
        try {
            const { sku } = req.params;
            const updateInventoryRequest = req.body;
            this.logRequest(req, `Updating inventory quantity for SKU: ${sku}`);
            
            if (!this.isValidSkuCode(sku)) {
                return this.createValidationErrorResponse(res, 'SKU', 'must be between 2 and 50 characters');
            }

            if (updateInventoryRequest.quantity === undefined || updateInventoryRequest.quantity < 0) {
                return this.sendBadRequestResponse(res, 'Quantity must be a non-negative number');
            }

            const result = await inventoryService.updateInventoryQuantity(sku, updateInventoryRequest);
            
            // If the service returns the standard format, pass it through
            if (result && result.status !== undefined && result.data !== undefined) {
                return res.status(result.status).json(result);
            }
            
            return this.sendSuccessResponse(res, result, `Inventory quantity updated successfully for SKU: ${sku}`);
        } catch (error) {
            console.error('[InventoryController] Error in updateInventoryQuantity:', error.message);
            return this.handleAxiosError(res, error, `Error updating inventory for SKU: ${req.params.sku}`);
        }
    }

    /**
     * Check if inventory exists for a SKU
     * GET /api/inventory/:sku/exists
     */
    async checkInventoryExists(req, res) {
        try {
            const { sku } = req.params;
            this.logRequest(req, `Checking inventory existence for SKU: ${sku}`);
            
            if (!this.isValidSkuCode(sku)) {
                return this.createValidationErrorResponse(res, 'SKU', 'must be between 2 and 50 characters');
            }

            const result = await inventoryService.checkInventoryExists(sku);
            
            // If the service returns the standard format, pass it through
            if (result && result.status !== undefined && result.data !== undefined) {
                return res.status(result.status).json(result);
            }
            
            return this.sendSuccessResponse(res, result, `Inventory existence check completed for SKU: ${sku}`);
        } catch (error) {
            console.error('[InventoryController] Error in checkInventoryExists:', error.message);
            return this.handleAxiosError(res, error, `Error checking inventory existence for SKU: ${req.params.sku}`);
        }
    }

    /**
     * Get all inventories
     * GET /api/inventory
     */
    async getAllInventories(req, res) {
        try {
            this.logRequest(req, 'Getting all inventories');
            
            const result = await inventoryService.getAllInventories();
            
            // If the service returns the standard format, pass it through
            if (result && result.status !== undefined && result.data !== undefined) {
                return res.status(result.status).json(result);
            }
            
            return this.sendSuccessResponse(res, result, 'All inventories retrieved successfully');
        } catch (error) {
            console.error('[InventoryController] Error in getAllInventories:', error.message);
            return this.handleAxiosError(res, error, 'Error retrieving inventories');
        }
    }

    /**
     * Health check - proxy to inventory service health check
     */
    async healthCheck(req, res) {
        try {
            console.log('[InventoryController] Performing health check');
            const result = await inventoryService.healthCheck();
            
            return this.sendSuccessResponse(res, result, 'Inventory service is healthy');
        } catch (error) {
            console.error('[InventoryController] Health check failed:', error.message);
            return this.sendServiceUnavailableResponse(res, 'Inventory service is unhealthy');
        }
    }
}

// Create an instance and export the methods
const inventoryController = new InventoryController();

module.exports = {
    createInventoryForProduct: (req, res) => inventoryController.createInventoryForProduct(req, res),
    getInventoryBySku: (req, res) => inventoryController.getInventoryBySku(req, res),
    updateInventoryQuantity: (req, res) => inventoryController.updateInventoryQuantity(req, res),
    checkInventoryExists: (req, res) => inventoryController.checkInventoryExists(req, res),
    getAllInventories: (req, res) => inventoryController.getAllInventories(req, res),
    healthCheck: (req, res) => inventoryController.healthCheck(req, res)
};
